src/spinnaker.cpp

```diff
@@ -0,0 +1,1049 @@
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return 0;
+}
+
+enum DeviceStatusCode
+spinnakercam_describe(const struct Driver* self_,
+                      struct DeviceIdentifier* identifier,
+                      uint64_t i)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerDriver*)self_)->describe(identifier, i);
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_open(struct Driver* self_, uint64_t device_id, struct Device** out)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerDriver*)self_)->open(device_id, out);
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_close(struct Driver* self_, struct Device* in)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerDriver*)self_)->close(in);
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_shutdown(struct Driver* self_)
+{
+    try {
+        CHECK(self_);
+        delete (struct SpinnakerDriver*)self_;
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+//
+// Driver implementation
+//
+
+SpinnakerDriver::SpinnakerDriver()
+  : Driver{
+      .device_count = ::spinnakercam_device_count,
+      .describe = ::spinnakercam_describe,
+      .open = ::spinnakercam_open,
+      .close = ::spinnakercam_close,
+      .shutdown = ::spinnakercam_shutdown,
+  },
+  system_(Spinnaker::System::GetInstance())
+{
+}
+
+SpinnakerDriver::~SpinnakerDriver()
+{
+    try {
+        shutdown();
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+}
+
+void
+SpinnakerDriver::describe(DeviceIdentifier* identifier, uint64_t i)
+{
+    check_spinnaker_camera_id(i);
+
+    Spinnaker::CameraList camera_list = system_->GetCameras();
+    Spinnaker::CameraPtr camera = camera_list.GetByIndex((unsigned int)i);
+    Spinnaker::GenApi::INodeMap& nodeMap = camera->GetTLDeviceNodeMap();
+
+    const Spinnaker::GenApi::CStringPtr vendor_name =
+      nodeMap.GetNode("DeviceVendorName");
+    const Spinnaker::GenApi::CStringPtr device_name =
+      nodeMap.GetNode("DeviceModelName");
+    const Spinnaker::GenApi::CStringPtr device_sn =
+      nodeMap.GetNode("DeviceSerialNumber");
```

Comment:
no change needed. But fwiw `auto` is ok.

Reply:
With Spinnaker, I prefer being more explicit about the types, so going to leave this.